### PR TITLE
Add truncindex to stores, to allow use of truncated IDs

### PR DESCRIPTION
**Pull request by nalind**
Add a truncindex to the layer, image, and container stores, so that layers, images, and containers can also be referred to by truncated versions of their IDs in any place where we previously accepted names.
    
For containers, we don't add a truncated index for referring to containers by layer ID, because there doesn't appear to be a minimum length in truncindex, and we don't want to mistakenly treat a truncated but not sufficiently-unique container ID as a layer ID, which would likely resolve to some other different container.

### CODE REVIEW

File: storage/layers.go [hunk 1]
```diff
@@ -467,6 +481,8 @@ func (r *layerStore) Unmount(id string) error {
 	}
 	if layer, ok := r.byname[id]; ok {
 		id = layer.ID
+	} else if longid, err := r.idindex.Get(id); err == nil {
```
Comment (runcom) [inline storage/layers.go]: can you have a function which fisrt check the `r.byname[id]` and the the truncindex to not repeat this else if over? skip this if you think it's not worth.

Comment (nalind) [inline storage/layers.go]: Might as well; it should make a dent in the lint warnings.

**Comment by nalind** [comment]
Okay, updated.  I think it got rid of a couple of lint warnings, at least.

**Comment by nalind** [comment]
Rebased on top of #24.

**Comment by rhatdan** [comment]
@runcom PTAL  We need to get this in.

**Comment by rhatdan** [comment]
:+1:

**Comment by runcom** [comment]
LGTM

**Comment by rhatdan** [comment]
@nalind Can we merge this now?

**Comment by nalind** [comment]
Sure, merging.

**Pull request merged**